### ISSUE TITLE
"Migrations" -- cb is not a function - HdWalletProvider

### ISSUE THREAD

**Issue by TamirCohen** [trufflesuite/truffle#4676]
- [x] I've [opened a support ticket](https://trufflesuite.zendesk.com/hc/en-us/requests/new) before filing this issue.

---------------------------

## Issue

Exception when migrating to ropsten

## Steps to Reproduce
`truffle migrate --network ropsten`
HardwareWallet-Provider version: 2.0.2 (Not Reprducing On 2.0.0)
truffle-config.js (Only the ropsten network config)
```
ropsten: {
      provider: function() {
        return new HDWalletProvider(secret.mnemonic, "wss://ropsten.infura.io/ws/v3/<API_KEY>")
      },
      network_id: 3,
      gas: 4000000
    }
```

### skipping compiling as expected
```
Compiling your contracts...
===========================
Everything is up to date, there is nothing to compile.
                                             
WARNING: Ganache forking only supports EIP-1193-compliant providers. Legacy support for send is currently enabled, but will be removed in a future version _without_ a breaking change
. To remove this warning, switch to an EIP-1193 provider. This error is probably caused by an old version of Web3's HttpProvider (or ganache < v7)

eth_getBlockByNumber

Migrations dry-run (simulation)
===============================
> Network name:    'ropsten-fork'
> Network id:      3
> Block gas limit: 8000000 (0x7a1200)


1_initial_migration.js
My comment: Initial migration works

```
### Error in the migration
```
Starting migrations...                                                                                                                                                                
======================                                                                                                                                                                
> Network name:    'ropsten'                                                                                                                                                          
> Network id:      3                                                                                                                                                                  
> Block gas limit: 8000000 (0x7a1200)


1_initial_migration.js
======================

   Deploying 'Migrations'
   ----------------------


Error:  *** Deployment Failed ***

"Migrations" -- cb is not a function.
```
### Stack trace
 ```
   at /home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/deployer/src/deployment.js:365:1
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Migration._deploy (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:70:1)
    at Migration._load (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:56:1)
    at Migration.run (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:217:1)
    at Object.runMigrations (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:150:1)
    at Object.runFrom (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:110:1)
    at Object.run (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:87:1)
    at runMigrations (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:76:1)
    at Object.module.exports [as run] (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:42:1)
    at Command.run (/home/user/.nvm/versions/node/v16.13.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/command.js:189:1)
```
## Environment

* Operating System: WSL2, Ubuntu 20.04.1 LTS
* Ethereum client: infura.io
* Truffle version (`truffle version`): 
```
Truffle v5.4.30 (core: 5.4.30)
Solidity - 0.8.11 (solc-js)
Node v16.13.2
Web3.js v1.5.3
```
* node version (`node --version`): v16.13.2
* npm version (`npm --version`): 8.4.0
* HardwareWallet-Provider version: 2.0.2

**Comment by HankHoneymoon** [comment]
I'm not sure if this is helpful or not, but several people have also filed the same issue on StackOverflow within the past 24 hours.
https://stackoverflow.com/questions/70911092/truffle-migrations-cb-is-not-a-function/70916931#70916931

**Comment by HankHoneymoon** [comment]
Right, workaround: I rolled @HDWalletProvider back to v2.0.1 and was able to migrate.

I assume there must be an issue with the new version for ppl who updated today.

**Comment by salahgo** [comment]
I have the same issue, probably related with the version 2.0.1+ of hdwallet-provider.
I suggest using 2.0.0 until they fix it: npm install --save @truffle/hdwallet-provider@2.0.0

**Comment by cds-amal** [comment]
Thanks for reporting this @TamirCohen! We'll try to get to the bottom of this! Appreciate the workaround @HankHoneymoon

**Comment by cds-amal** [comment]
I was able to reproduce this w/ instrumentation (`DEBUG=contract:*`)

``` shell
1_initial_migration.js
======================

   Deploying 'Migrations'
   ----------------------
  contract:promievent hijacking! +0ms
  contract:promievent hash: undefined +1ms
  contract:promievent e.stack: TypeError: cb is not a function
    at HookedWalletSubprovider.signTransaction (/home/amal/.scratch/2022/01/30/4676-hdwallet/mc/node_modules/@truffle/hdwallet-provider/src/index.ts:199:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) +1ms
  contract:promievent originalStackTrace: Error:
    at new PromiEvent (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/contract/lib/promievent.js:8:1)
    at /home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/contract/lib/execute.js:253:1
    at Function.new (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/contract/lib/contract/constructorMethods.js:57:1)
    at /home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/deployer/src/deployment.js:335:1
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at Migration._deploy (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:70:1)
    at Migration._load (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:56:1)
    at Migration.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:217:1)
    at Object.runMigrations (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:150:1)
    at Object.runFrom (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:110:1)
    at Object.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:87:1)
    at runMigrations (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:76:1)
    at Object.module.exports [as run] (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:42:1)
    at Command.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/command.js:189:1) +0ms
  contract:promievent solidityStackTrace: undefined +0ms


Error:  *** Deployment Failed ***

"Migrations" -- cb is not a function.

    at /home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/deployer/src/deployment.js:365:1
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at Migration._deploy (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:70:1)
    at Migration._load (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:56:1)
    at Migration.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:217:1)
    at Object.runMigrations (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:150:1)
    at Object.runFrom (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:110:1)
    at Object.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/migrate/index.js:87:1)
    at runMigrations (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:76:1)
    at Object.module.exports [as run] (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/commands/migrate/run.js:42:1)
    at Command.run (/home/amal/.nvm/versions/node/v14.18.2/lib/node_modules/truffle/build/webpack:/packages/core/lib/command.js:189:1)
Truffle v5.4.30 (core: 5.4.30)
Node v14.18.2
```

**Comment by teja2002** [comment]
> 

It resolves my problem. Thank you for suggesting that

**Comment by joaneeet7** [comment]
Use    "@truffle/hdwallet-provider": "2.0.2", in package.json.
Then use var HDWalletProvider = require("@truffle/hdwallet-provider") in your truffle-config

**Comment by cds-amal** [comment]
This should be resolved with the latest [truffle](https://github.com/trufflesuite/truffle/releases/tag/v5.4.32).
